File: src/main/java/smart4aviation/entity/FlightCargo.java
package smart4aviation.entity;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.List;

/**
 * I renamed the entity into FlightCargo to avoid misunderstanding when considering the Cargo elements
 */
@JsonSerialize
@AllArgsConstructor
@NoArgsConstructor
@Getter
public class FlightCargo {

    private final Double LBS_TO_KG = 1 / 2.20462;

    private Long flightId;
    private List<TransportUnit> baggage;
    private List<TransportUnit> cargo;

    public Integer getBaggageTotalPieces() {
        return baggage.stream().map(TransportUnit::getPieces).reduce(0, Integer::sum);
    }

    public Double getBaggageTotalWeight() {
        return baggage.stream().map(baggage -> baggage.getWeight() * getCOMConversionRatio(baggage.getWeightUnit())).reduce(.0, Double::sum);
    }

    public Integer getCargoTotalPieces() {
        return cargo.stream().map(TransportUnit::getPieces).reduce(0, Integer::sum);
    }

    public Double getCargoTotalWeight() {
        return cargo.stream().map(cargo -> cargo.getWeight() * getCOMConversionRatio(cargo.getWeightUnit())).reduce(.0, Double::sum);
    }

    private Double getCOMConversionRatio(String UOM) {
        return switch (UOM) {
            case "kg" -> 1.0;
            case "lb" -> LBS_TO_KG;
            default -> .0;
        };
    }
}


@JsonSerialize
@AllArgsConstructor
@NoArgsConstructor
@Getter
class TransportUnit {
    private Long id;
    private Integer weight;
    private String weightUnit;
    private Integer pieces;
}

File: src/main/kotlin/com/example/smart4aviation/entity/FlightCargo.kt
package com.example.smart4aviation.entity

import com.fasterxml.jackson.databind.annotation.JsonSerialize

/**
 * I renamed the entity into FlightCargo to avoid misunderstanding when considering the Cargo elements
 */
@JsonSerialize
class FlightCargo(
    val flightId: Long,
    val baggage: TransportUnitList,
    val cargo: TransportUnitList
) {}

@JsonSerialize
class TransportUnitList() : MutableList<TransportUnit> by mutableListOf() {

    private val LBS_TO_KG = 1 / 2.20462

    fun getTotalPieces(): Int {
        return this.sumOf { it.pieces }
    }

    fun getTotalWeight(): Double {
        return this.sumOf { it.weight * getCOMConversionRatio(it.weightUnit) }
    }

    private fun getCOMConversionRatio(UOM: String): Double {
        return when (UOM) {
            "kg" -> 1.0
            "lb" -> LBS_TO_KG
            else -> .0
        }

    }
}

@JsonSerialize
class TransportUnit(
    val id: Long,
    val weight: Int,
    val weightUnit: String,
    val pieces: Int
) {}